### PR TITLE
#3133 - Build Student Unmet Need Report: BC Public Institutions (Part 2)

### PR DESCRIPTION
**Pull request by sh16011993**
### As a part of this PR, the following were completed:

**- Wrote the following e2e test:**

ReportInstitutionsController(e2e)-exportReport
    √ Should generate the Student Unmet Need Report when a report generation request is made with the appropriate filters.

### Screenshots:

image

--------------------------------------------

**- Updated the Report SQL to map the Application Disability Status column as per the requirement.**

```SQL
SELECT
      users.first_name AS "Student First Name",
      users.last_name AS "Student Last Name",
      sin_validations.sin AS "SIN",
      applications.student_number AS "Student Number",
      users.email AS "Student Email Address",
      students.contact_info ->> 'phone' AS "Student Phone Number",
      institution_locations.institution_code AS "Institution Location Code",
      institution_locations.name AS "Institution Location Name",
      applications.application_number AS "Application Number",
      CAST(
        CAST(student_assessments.assessment_date AS date) AS varchar
      ) AS "Assessment Date",
      education_programs_offerings.offering_intensity AS "Study Intensity (PT or FT)",
      students.disability_status AS "Profile Disability Status",
      CASE
        WHEN student_assessments.workflow_data -> 'calculatedData' ->> 'pdppdStatus' = 'false' THEN 'no'
        WHEN student_assessments.workflow_data -> 'calculatedData' ->> 'pdppdStatus' = 'true' THEN 'yes'
      END AS "Application Disability Status",
      CAST(
        education_programs_offerings.study_start_date AS varchar
      ) AS "Study Start Date",
      CAST(
        education_programs_offerings.study_end_date AS varchar
      ) AS "Study End Date",
      education_programs.program_name AS "Program Name",
      education_programs.credential_type AS "Program Credential Type",
      education_programs.cip_code AS "CIP Code",
      education_programs.completion_years AS "Program Length",
      education_programs.sabc_code AS "SABC Program Code",
      education_programs_offerings.offering_name AS "Offering Name",
      education_programs_offerings.year_of_study AS "Year of Study",
      applications.data ->> 'indigenousStatus' AS "Indigenous person status",
      applications.data ->> 'citizenship' AS "Citizenship Status",
      applications.data ->> 'youthInCare' AS "Youth in Care Flag",
      applications.data ->> 'custodyOfChildWelfare' AS "Youth in Care beyond age 19",
      applications.relationship_status AS "Marital Status",
      applications.data ->> 'dependantstatus' AS "Independant/Dependant",
      student_assessments.workflow_data -> 'calculatedData' ->> 'totalEligibleDependents' AS "Number of Eligible Dependants Total",
      CASE
        WHEN education_programs_offerings.offering_intensity = 'Part Time' THEN student_assessments.assessment_data ->> 'totalAssessmentNeed'
        WHEN education_programs_offerings.offering_intensity = 'Full Time' THEN student_assessments.assessment_data ->> 'totalAssessedCost'
      END AS "Federal/Provincial Assessed Costs",
      student_assessments.assessment_data ->> 'totalFederalAssessedResources' AS "Federal Assessed Resources",
      student_assessments.assessment_data ->> 'federalAssessmentNeed' AS "Federal assessed need",
      student_assessments.assessment_data ->> 'totalProvincialAssessedResources' AS "Provincial Assessed Resources",
      student_assessments.assessment_data ->> 'provincialAssessmentNeed' AS "Provincial assessed need",
      student_assessments.assessment_data ->> 'finalAwardTotal' AS "Total assistance",
      student_assessments.assessment_data ->> 'finalFederalAwardNetCSGPAmount' AS "Estimated CSGP",
      student_assessments.assessment_data ->> 'finalFederalAwardNetCSPTAmount' AS "Estimated CSPT",
      student_assessments.assessment_data ->> 'finalFederalAwardNetCSGDAmount' AS "Estimated CSGD",
      student_assessments.assessment_data ->> 'finalProvincialAwardNetBCAGAmount' AS "Estimated BCAG",
      student_assessments.assessment_data ->> 'finalProvincialAwardNetSBSDAmount' AS "Estimated SBSD",
      student_assessments.assessment_data ->> 'finalFederalAwardNetCSLPAmount' AS "Estimated CSLP",
      student_assessments.assessment_data ->> 'finalProvincialAwardNetBCSLAmount' AS "Estimated BCSL",
      student_assessments.assessment_data ->> 'finalFederalAwardNetCSGFAmount' AS "Estimated CSGF",
      student_assessments.assessment_data ->> 'finalProvincialAwardNetBGPDAmount' AS "Estimated BGPD"
    FROM
      sims.applications applications
      INNER JOIN sims.students students ON students.id = applications.student_id
      INNER JOIN sims.sin_validations sin_validations ON sin_validations.student_id = students.id
      INNER JOIN sims.users users ON users.id = students.user_id
      INNER JOIN sims.student_assessments student_assessments ON student_assessments.id = applications.current_assessment_id
      INNER JOIN sims.education_programs_offerings education_programs_offerings ON education_programs_offerings.id = student_assessments.offering_id
      INNER JOIN sims.education_programs education_programs ON education_programs.id = education_programs_offerings.program_id
      INNER JOIN sims.institution_locations institution_locations ON institution_locations.id = applications.location_id
    WHERE
      applications.application_status IN ('Assessment', 'Enrolment', 'Completed')
      AND applications.is_archived = FALSE
      AND education_programs_offerings.offering_intensity = ANY(:offeringIntensity)
      AND applications.program_year_id = :programYear
      AND institution_locations.institution_id = :institutionId;
```

**Updated Report Screenshots:**

**1st Half:**

image

**2nd Half:**

image

--------------------------------------------------

**Bug: Institution Offering Details Report has an Unexpected Error**

image

**Fix: Updated the formio definition to include the missing condition to conditionally display the Offering Intensity Filter.** **This fixed the issue and the report is generated as below:**

image

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/report/_tests_/e2e/report.institutions.controller.exportReport.e2e-spec.ts [hunk 1]
```diff
@@ -255,4 +260,209 @@ describe("ReportInstitutionsController(e2e)-exportReport", () => {
...
+        const parsedResult = parse(fileContent, {
+          header: true,
+        });
+        expect(parsedResult.data).toEqual(
+          expect.arrayContaining([
+            {
+              "Application Disability Status": "no",
+              "Application Number": savedApplication.applicationNumber,
+              "Assessment Date": getISODateOnlyString(
+                savedApplication.currentAssessment.assessmentDate,
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/report/_tests_/e2e/report.institutions.controller.exportReport.e2e-spec.ts:346]: please create contants for assessmentDate and currentAssessment

File: sources/packages/backend/libs/test-utils/src/factories/application.ts [hunk 2]
```diff
@@ -101,9 +103,11 @@ export async function saveFakeApplicationDisbursements(
     student?: Student;
     msfaaNumber?: MSFAANumber;
     program?: EducationProgram;
+    programYear?: ProgramYear;
```
Comment (guru-aot) [inline sources/packages/backend/libs/test-utils/src/factories/application.ts:106]: can you please explain where are we using this in the code?

File: sources/packages/backend/apps/api/src/route-controllers/report/_tests_/e2e/report.institutions.controller.exportReport.e2e-spec.ts [hunk 3]
```diff
@@ -255,4 +260,209 @@ describe("ReportInstitutionsController(e2e)-exportReport", () => {
...
+              ),
+              "Estimated CSGD": String(
+                savedApplication.currentAssessment.assessmentData
+                  .finalFederalAwardNetCSGDAmount,
+              ),
+              "Estimated CSGF": String(
+                savedApplication.currentAssessment.assessmentData[
+                  "finalFederalAwardNetCSGFAmount"
+                ],
+              ),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/report/_tests_/e2e/report.institutions.controller.exportReport.e2e-spec.ts:375]: The `assessment` is defined as 
```ts
export type Assessment = FullTimeAssessment | PartTimeAssessment;
```
and the indentation is to have the cast executed when needed.
So, if we know those values are part of a full-time assessment, casting to `FullTimeAssessment` will allow typed access to the properties.

File: sources/packages/backend/apps/api/src/route-controllers/report/_tests_/e2e/report.institutions.controller.exportReport.e2e-spec.ts [hunk 4]
```diff
@@ -255,4 +260,209 @@ describe("ReportInstitutionsController(e2e)-exportReport", () => {
...
+              "Application Number": savedApplication.applicationNumber,
+              "Assessment Date": getISODateOnlyString(
+                savedApplication.currentAssessment.assessmentDate,
+              ),
+              "CIP Code":
+                savedApplication.currentAssessment.offering.educationProgram
+                  .cipCode,
+              "Citizenship Status":
+                savedApplication.currentAssessment.application.data.citizenship,
+              "Estimated BCAG": String(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/report/_tests_/e2e/report.institutions.controller.exportReport.e2e-spec.ts:353]: Not a blocker but what I believe to be more common in the source code is the use of the `.toString`.

File: sources/packages/backend/libs/test-utils/src/factories/application.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/test-utils/src/factories/application.ts:106]: The entire `relations` is passed to the `saveFakeApplication` which can potentially use it.

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, please take a look at the comments.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, looks good 👍

**Pull request merged**